mm2src/coins/tx_history_storage/mod.rs

```diff
@@ -0,0 +1,183 @@
+pub fn token_id_from_tx_type(tx_type: &TransactionType) -> String {
+    match tx_type {
+        TransactionType::TokenTransfer(token_id) => format!("{:02x}", token_id),
+        _ => String::new(),
+    }
+}
+
+#[derive(Debug, Display)]
+pub enum CreateTxHistoryStorageError {
+    Internal(String),
+}
+
+/// `TxHistoryStorageBuilder` is used to create an instance that implements the `TxHistoryStorage` trait.
+pub struct TxHistoryStorageBuilder<'a> {
+    ctx: &'a MmArc,
+}
+
+impl<'a> TxHistoryStorageBuilder<'a> {
+    #[inline]
+    pub fn new(ctx: &MmArc) -> TxHistoryStorageBuilder<'_> { TxHistoryStorageBuilder { ctx } }
+
+    #[inline]
+    pub fn build(self) -> MmResult<impl TxHistoryStorage, CreateTxHistoryStorageError> {
+        #[cfg(target_arch = "wasm32")]
+        return wasm::IndexedDbTxHistoryStorage::new(self.ctx);
+        #[cfg(not(target_arch = "wasm32"))]
+        sql_tx_history_storage_v2::SqliteTxHistoryStorage::new(self.ctx)
+    }
+}
+
+/// Whether transaction is unconfirmed or confirmed.
+/// Serializes to either `0u8` or `1u8` correspondingly.
+#[repr(u8)]
+#[derive(Clone, Copy, Debug)]
+pub enum ConfirmationStatus {
+    Unconfirmed = 0,
+    Confirmed = 1,
+}
+
+impl ConfirmationStatus {
+    #[inline]
+    pub fn from_block_height<Height: Zero>(height: Height) -> ConfirmationStatus {
+        if height.is_zero() {
+            ConfirmationStatus::Unconfirmed
+        } else {
+            ConfirmationStatus::Confirmed
+        }
+    }
+}
+
+impl Serialize for ConfirmationStatus {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        (*self as u8).serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for ConfirmationStatus {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let code = u8::deserialize(deserializer)?;
+        match code {
+            0 => Ok(ConfirmationStatus::Unconfirmed),
+            1 => Ok(ConfirmationStatus::Confirmed),
+            unknown => Err(D::Error::custom(format!(
+                "Expected either '0' or '1' confirmation status, found '{}'",
+                unknown
+            ))),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct WalletId {
+    ticker: String,
+    hd_wallet_rmd160: Option<H160>,
+}
+
+impl WalletId {
+    #[inline]
+    pub fn new(ticker: String) -> WalletId {
+        WalletId {
+            ticker,
+            hd_wallet_rmd160: None,
+        }
+    }
+
+    #[inline]
+    pub fn set_hd_wallet_rmd160(&mut self, hd_wallet_rmd160: H160) { self.hd_wallet_rmd160 = Some(hd_wallet_rmd160); }
+
+    #[inline]
+    pub fn with_hd_wallet_rmd160(mut self, hd_wallet_rmd160: H160) -> WalletId {
+        self.set_hd_wallet_rmd160(hd_wallet_rmd160);
+        self
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct GetTxHistoryFilters {
+    token_id: Option<String>,
+    for_addresses: Option<FilteringAddresses>,
```

Comment:
Should the address maybe be the key/name for a separate table? Can it make some queries more optimal?

Reply:
I'm not sure if I understand the question correctly.
Do you mean name the table, for example, `KMD_RJTYiYeJ8eVvJ53n2YbrVmxWNNMVZjDGLh_RHzSYSHv3G6J8xL3MyGH3y2gU588VCTC7X` if the user enabled two HD addresses `RJTYiYeJ8eVvJ53n2YbrVmxWNNMVZjDGLh` and `RHzSYSHv3G6J8xL3MyGH3y2gU588VCTC7X`?

But if the user enables a new address `RRnMcSeKiLrNdbp91qNVQwwXx5azD4S4CD`, we get a new table that needs to be fulfilled with the same transactions (for `RJTYiYeJ8eVvJ53n2YbrVmxWNNMVZjDGLh` and `RHzSYSHv3G6J8xL3MyGH3y2gU588VCTC7X`) and other transactions for the `RRnMcSeKiLrNdbp91qNVQwwXx5azD4S4CD` address.

Reply:
I don't mean creating the table for multiple addresses, I mean `KMD_RJTYiYeJ8eVvJ53n2YbrVmxWNNMVZjDGLh`, `KMD_RHzSYSHv3G6J8xL3MyGH3y2gU588VCTC7X`, `KMD_RRnMcSeKiLrNdbp91qNVQwwXx5azD4S4CD` and so on.

Reply:
I’ve just realized that we will have a big trouble with such DB structure, if we need to query history for multiple addresses at once, especially with filtering and pagination. Resolving this as not actual anymore :)